aptos-move/aptos-transaction-simulation/src/state_store.rs

```diff
@@ -0,0 +1,447 @@
+            Self::Left(l) => l.id(),
+            Self::Right(r) => r.id(),
+        }
+    }
+
+    fn get_state_value(&self, state_key: &Self::Key) -> StateViewResult<Option<StateValue>> {
+        match self {
+            Self::Left(l) => l.get_state_value(state_key),
+            Self::Right(r) => r.get_state_value(state_key),
+        }
+    }
+
+    fn get_usage(&self) -> StateViewResult<StateStorageUsage> {
+        match self {
+            Self::Left(l) => l.get_usage(),
+            Self::Right(r) => r.get_usage(),
+        }
+    }
+
+    fn contains_state(&self, state_key: &Self::Key) -> StateViewResult<bool> {
+        match self {
+            Self::Left(l) => l.contains_state(state_key),
+            Self::Right(r) => r.contains_state(state_key),
+        }
+    }
+}
+
+/***************************************************************************************************
+ * Delta State Store
+ *
+ **************************************************************************************************/
+/// A state storage that allows changes to be stacked on top of a base state view.
+///
+/// This is useful for staging reversible state changes or performing simulations on top of
+/// remote states.
+pub struct DeltaStateStore<V> {
+    base: V,
+    states: Mutex<HashMap<StateKey, Option<StateValue>>>,
+}
+
+impl<V> TStateView for DeltaStateStore<V>
+where
+    V: TStateView<Key = StateKey>,
+{
+    type Key = StateKey;
+
+    // TODO: fn id()?
+
+    fn get_state_value(&self, state_key: &Self::Key) -> StateViewResult<Option<StateValue>> {
+        if let Some(res) = self.states.lock().get(state_key) {
+            return Ok(res.clone());
+        }
+        self.base.get_state_value(state_key)
+    }
+
+    fn get_usage(&self) -> StateViewResult<StateStorageUsage> {
+        Ok(StateStorageUsage::Untracked)
+    }
+
+    fn contains_state(&self, state_key: &Self::Key) -> StateViewResult<bool> {
+        let states = self.states.lock();
+
+        match states.get(state_key) {
+            Some(Some(_)) => return Ok(true),
+            Some(None) => return Ok(false),
+            None => (),
+        }
+
+        _ = states; // drop the lock as early as possible
+
+        self.base.contains_state(state_key)
+    }
+}
+
+impl<V> SimulationStateStore for DeltaStateStore<V>
+where
+    V: TStateView<Key = StateKey>,
+{
+    fn set_state_value(&self, state_key: StateKey, state_val: StateValue) -> Result<()> {
+        self.states.lock().insert(state_key, Some(state_val));
+        Ok(())
+    }
+
+    fn remove_state_value(&self, state_key: &StateKey) -> Result<()> {
+        self.states.lock().remove(state_key);
+        Ok(())
+    }
+
+    fn apply_write_set(&self, write_set: &WriteSet) -> Result<()> {
+        let mut states = self.states.lock();
+
+        for (state_key, write_op) in write_set {
+            match write_op.as_state_value() {
+                None => {
+                    states.remove(state_key);
```

Comment:
This seems wrong. This happens when the write is a deletion so you should keep `None` in states, otherwise why states store `Option<StateValue>`?

PoC: if you currently store a value in states, and you have a deletion, you will remove it but the base may contain the original version which you would now return on read, which is wrong (you should return None from read)

Reply:
Good catch. I fixed `fn remove_state_value` fixed in my local version but didn't catch this one.

Reply:
Right, `remove_state_value` also should be setting the value to None if exists